presto-main/src/main/java/io/prestosql/operator/aggregation/ApproximateMostFrequentHistogram.java

```diff
@@ -0,0 +1,147 @@
+ *  as described in:
+ *  <i>Efficient Computation of Frequent and Top-k Elements in Data Streams</i>
+ *  by Metwally, Agrawal, and Abbadi
+ * @param <K>
+ */
+public class ApproximateMostFrequentHistogram<K>
+{
+    private static final byte FORMAT_TAG = 0;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(ApproximateMostFrequentHistogram.class).instanceSize();
+
+    private StreamSummary<K> streamSummary;
+    private final int maxBuckets;
+    private final int capacity;
+    private final ApproximateMostFrequentBucketSerializer<K> serializer;
+    private final ApproximateMostFrequentBucketDeserializer<K> deserializer;
+
+    /**
+     * @param maxBuckets The maximum number of elements stored in the bucket.
+     * @param capacity The maximum capacity of the stream summary data structure.
+     * @param serializer It serializes a bucket into varbinary slice.
+     * @param deserializer It appends a bucket into the histogram.
+     */
+    public ApproximateMostFrequentHistogram(int maxBuckets,
+                                            int capacity,
+                                            ApproximateMostFrequentBucketSerializer<K> serializer,
+                                            ApproximateMostFrequentBucketDeserializer<K> deserializer)
+    {
+        requireNonNull(serializer, "serializer is null");
+        requireNonNull(deserializer, "deserializer is null");
+        streamSummary = new StreamSummary<>(capacity);
+        this.maxBuckets = maxBuckets;
+        this.capacity = capacity;
+        this.serializer = serializer;
+        this.deserializer = deserializer;
+    }
+
+    public ApproximateMostFrequentHistogram(Slice serialized,
+                                            ApproximateMostFrequentBucketSerializer<K> serializer,
+                                            ApproximateMostFrequentBucketDeserializer<K> deserializer)
+    {
+        SliceInput input = serialized.getInput();
+
+        checkArgument(input.readByte() == FORMAT_TAG, "Unsupported format tag");
+
+        this.maxBuckets = input.readInt();
+        this.capacity = input.readInt();
+        int bucketSize = input.readInt();
+        this.streamSummary = new StreamSummary<>(capacity);
+        this.serializer = serializer;
+        this.deserializer = deserializer;
+
+        for (int i = 0; i < bucketSize; i++) {
+            this.deserializer.deserialize(input, this);
+        }
+    }
+
+    public void add(K value)
+    {
+        streamSummary.offer(value);
+    }
+
+    public void add(K value, long incrementCount)
+    {
+        streamSummary.offer(value, toIntExact(incrementCount));
+    }
+
+    public Slice serialize()
+    {
+        List<Counter<K>> counters = streamSummary.topK(maxBuckets);
+        int estimatedSliceSize = Byte.BYTES + counters.size() * Integer.BYTES * 3; // FORMAT_TAG + Bytes allocated for total buckets
+        DynamicSliceOutput output = new DynamicSliceOutput(estimatedSliceSize * 3);
```

Comment:
This calculation seems off. From the code below, it needs:
* 1 byte for the format tag
* 4 bytes for maxBuckets 
* 4 bytes for capacity
* 4 bytes for the number of counters
* 8 * counters.size() bytes for the count
* for long keys, 8 * counters.size() bytes for the keys. For string keys, it can vary, but if we just pick 8, it would be an ok estimate that works for most cases (add a comment explaining why we're picking such a number)